### PR TITLE
Forex Typelevel

### PR DESCRIPTION
**Pull request by gvolpe**
This a proposal to give our candidates the same `forex` exercise using the Typelevel stack instead of `eff`, `grafter`, etc.

It is by no mean a final version and I would really appreciate your comments and suggestions.

After looking at it with @yoohaemin we realized that it now looks easier so I wonder if we can add some extra difficulties. 

I'd expect the candidate to:

- Solve the error handling problem (right now there's a dummy ADT that doesn't say much).
- Write tests for the HTTP Routes (it is just a function in `http4s `) including json codecs.
- Write tests for the business logic `RatesProgram` (just adapting errors as of now).

### CODE REVIEW

File: forex-mtl/.scalafmt.conf [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+maxColumn                                    = 120
+newlines.afterImplicitKWInVerticalMultiline  = true
+newlines.beforeImplicitKWInVerticalMultiline = true
+project.excludeFilters                       = [ .scalafmt.conf ]
+project.git                                  = true
+rewrite.rules                                = [PreferCurlyFors, RedundantBraces, RedundantParens, SortImports]
+
+spaces.inImportCurlyBraces                   = true
+style                                        = defaultWithAlign
+unindentTopLevelOperators                    = true
```
Comment (abalonperin) [inline forex-mtl/.scalafmt.conf]: newline

File: forex-mtl/src/main/resources/application.conf [hunk 2]
```diff
@@ -0,0 +1,7 @@
+app {
+  http {
+    host = "0.0.0.0"
+    port = 8080
+    timeout = 40 seconds
+  }
+}
```
Comment (abalonperin) [inline forex-mtl/src/main/resources/application.conf]: newline

File: forex-mtl/src/main/resources/logback.xml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+            <pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n</pattern>
+        </encoder>
+    </appender>
+
+    <root level="INFO">
+        <appender-ref ref="STDOUT" />
+    </root>
+
+    <logger name="http4s"/>
+</configuration>
```
Comment (abalonperin) [inline forex-mtl/src/main/resources/logback.xml]: newline

File: forex-mtl/src/main/scala/forex/config/ApplicationConfig.scala [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+case class ApplicationConfig(
+    http: HttpConfig,
+)
+
+case class HttpConfig(
+    host: String,
+    port: Int,
+    timeout: FiniteDuration
+)
```
Comment (abalonperin) [inline forex-mtl/src/main/scala/forex/config/ApplicationConfig.scala]: newline

File: forex-mtl/src/main/scala/forex/http/package.scala [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  implicit def valueClassEncoder[A: UnwrappedEncoder]: Encoder[A] = implicitly
+  implicit def valueClassDecoder[A: UnwrappedDecoder]: Decoder[A] = implicitly
+
+  implicit def enumEncoder[A: EnumerationEncoder]: Encoder[A] = implicitly
+  implicit def enumDecoder[A: EnumerationDecoder]: Decoder[A] = implicitly
+
+  implicit def jsonDecoder[A <: Product: Decoder, F[_]: Sync]: EntityDecoder[F, A] = jsonOf[F, A]
+  implicit def jsonEncoder[A <: Product: Encoder, F[_]: Sync]: EntityEncoder[F, A] = jsonEncoderOf[F, A]
+
+}
```
Comment (abalonperin) [inline forex-mtl/src/main/scala/forex/http/package.scala]: newline

**Comment by abalonperin** [comment]
I like the simplicity. However, I wonder if we should keep the layers (api -> process -> services -> repositories) that we have in the non-mtl style interview and in most of our services.

**Comment by gvolpe** [comment]
@abalonperin we can do that if it's the general agreement. However let me explain why I don't sympathize with such names:

- `interfaces/api`: is used for all the `http` routes and `json` encoders, hence I prefer to call it `http`.
- `processes`: for the business logic. In tagless final I saw more and more people calling it `programs` (and I personally preferred this name) event though I can't say it's the standard.
- `services`: is used for interfaces that deal with remote http stuff, databases or something like that. I called `client` because I think it seems more appropriated for `OneForge` as it is an `http client` and secondly because the name `services` reminds me of Java services but this is personal preference.

**Comment by abalonperin** [comment]
@gvolpe I don't have a strong preference for either version. However, I believe that keeping a standard across projects and teams would help if we ever need to make changes in each other services. Maybe other people hove different opinions on this though.

How would you name things if you have to expose a repository? Would you wrap it inside a `client`? Or maybe call the repo directly in the process?

File: forex-mtl/src/main/scala/forex/rates/errors.scala [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package forex.rates
+
+object errors {
+
+  sealed trait RateError extends Exception
```
Comment (paulroseau) [inline forex-mtl/src/main/scala/forex/rates/errors.scala]: We used to organize this differently like : 
```
sealed trait RateError extends Exception
object RateError {
   final case class RemoteClient(msg: String) extends RateError
}
```
any reason why you change this ?

Comment (gvolpe) [reply forex-mtl/src/main/scala/forex/rates/errors.scala]: No reason, can do that :+1:

File: forex-mtl/project/Dependencies.scala [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    lazy val http4sCirce         = http4s("http4s-circe")
+    lazy val circeCore           = circe("circe-core")
+    lazy val circeGeneric        = circe("circe-generic")
+    lazy val circeGenericExt     = circe("circe-generic-extras")
+    lazy val circeParser         = circe("circe-parser")
+    lazy val circeJava8          = circe("circe-java8")
+    lazy val pureConfig          = "com.github.pureconfig" %% "pureconfig"                 % Versions.pureConfig
+
+    // Compiler plugins
+    lazy val betterMonadicFor    = "com.olegpy"            %% "better-monadic-for"         % Versions.betterMonadicFor
```
[the remark this answers is not in the document]
Comment (gvolpe) [reply forex-mtl/project/Dependencies.scala]: It's one of my compiler defaults, as important as `kind-projector` :)

Exactly what the name suggests, better monadic for's. Avoid the extra `map` calls and lets you do stuff like this `(a, b) <- monadicTuple` without having to define `withFilter`.

More info here: https://github.com/oleg-py/better-monadic-for

File: forex-mtl/src/main/resources/application.conf (hunk 2 above)
Comment (gvolpe) [reply forex-mtl/src/main/resources/application.conf]: @abalonperin not sure what you mean with all these newline comments, can you elaborate?

Comment (gvolpe) [reply forex-mtl/src/main/resources/application.conf]: Nevermind, I guess this is the answer: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

File: forex-mtl/src/main/scala/forex/programs/package.scala [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package forex
+
+import forex.domain._
+
+package object programs {
+  type RatesProgram[F[_]] = rates.Algebra[F]
+  final val RatesProgram = rates.Program
+
+
+  final case class GetRatesRequest(
```
[the remark this answers is not in the document]
Comment (abalonperin) [reply forex-mtl/src/main/scala/forex/programs/package.scala]: I have moved it in a previous commit maybe your view is outdated.

**Comment by gvolpe** [comment]
Bravo :tada: It took only half a year :smile: (/cc @pjan you let me down here!)

**Pull request merged**